modules/wmf/video_stream_wmf.cpp

```diff
@@ -0,0 +1,723 @@
+	SafeRelease(pOutType);
+	return S_OK;
+}
+
+HRESULT CreateSampleGrabber(UINT width, UINT height, SampleGrabberCallback *pSampleGrabber, IMFActivate **pSinkActivate) {
+	HRESULT hr = S_OK;
+	IMFMediaType *pType = NULL;
+
+	CHECK_HR(MFCreateMediaType(&pType));
+	CHECK_HR(pType->SetGUID(MF_MT_MAJOR_TYPE, MFMediaType_Video));
+	CHECK_HR(pType->SetGUID(MF_MT_SUBTYPE, MFVideoFormat_NV12));
+	CHECK_HR(MFSetAttributeSize(pType, MF_MT_FRAME_SIZE, width, height));
+	CHECK_HR(pType->SetUINT32(MF_MT_FIXED_SIZE_SAMPLES, TRUE));
+	CHECK_HR(pType->SetUINT32(MF_MT_ALL_SAMPLES_INDEPENDENT, TRUE));
+	CHECK_HR(MFSetAttributeRatio(pType, MF_MT_PIXEL_ASPECT_RATIO, 1, 1));
+	CHECK_HR(MFCreateSampleGrabberSinkActivate(pType, pSampleGrabber, pSinkActivate));
+
+	SafeRelease(pType);
+	return hr;
+}
+
+// Create the topology.
+HRESULT CreateTopology(IMFMediaSource *pSource, SampleGrabberCallback *pSampleGrabber, IMFTopology **ppTopo, VideoStreamPlaybackWMF::StreamInfo *info) {
+	IMFTopology *pTopology = NULL;
+	IMFPresentationDescriptor *pPD = NULL;
+	IMFStreamDescriptor *pSD = NULL;
+	IMFMediaTypeHandler *pHandler = NULL;
+	IMFTopologyNode *inputNode = NULL;
+	IMFTopologyNode *outputNode = NULL;
+	IMFTopologyNode *inputNodeAudio = NULL;
+	IMFTopologyNode *outputNodeAudio = NULL;
+	IMFActivate *audioActivate = NULL;
+
+	HRESULT hr = S_OK;
+
+	CHECK_HR(MFCreateTopology(&pTopology));
+	CHECK_HR(pSource->CreatePresentationDescriptor(&pPD));
+
+	DWORD cStreams = 0;
+	CHECK_HR(pPD->GetStreamDescriptorCount(&cStreams));
+
+	print_line(itos(cStreams) + " streams");
+
+	for (DWORD i = 0; i < cStreams; i++) {
+		BOOL bSelected = FALSE;
+		GUID majorType;
+
+		CHECK_HR(pPD->GetStreamDescriptorByIndex(i, &bSelected, &pSD));
+		CHECK_HR(pSD->GetMediaTypeHandler(&pHandler));
+		CHECK_HR(pHandler->GetMajorType(&majorType));
+
+		if (majorType == MFMediaType_Video && bSelected) {
+			print_line("Video Stream");
+
+			IMFMediaType *pType = NULL;
+			CHECK_HR(pHandler->GetMediaTypeByIndex(0, &pType));
+			UINT32 width = 0, height = 0;
+			MFGetAttributeSize(pType, MF_MT_FRAME_SIZE, &width, &height);
+
+			IMFActivate *pSinkActivate = NULL;
+			CHECK_HR(CreateSampleGrabber(width, height, pSampleGrabber, &pSinkActivate));
+			IMFTransform *pColorTransform = NULL;
+			CHECK_HR(AddColourConversionNode(pTopology, pType, &pColorTransform));
+			pSampleGrabber->set_color_transform(pColorTransform);
+
+			CHECK_HR(AddSourceNode(pTopology, pSource, pPD, pSD, &inputNode));
+			CHECK_HR(AddOutputNode(pTopology, pSinkActivate, 0, &outputNode));
+
+			CHECK_HR(inputNode->ConnectOutput(0, outputNode, 0));
+
+			info->size.x = width;
+			info->size.y = height;
+			print_line("Width & Height " + itos(width) + "x" + itos(height));
+
+			UINT32 numerator = 0, denominator = 1;
+			MFGetAttributeRatio(pType, MF_MT_FRAME_RATE, &numerator, &denominator);
+			if (denominator > 0) {
+				info->fps = (float)numerator / (float)denominator;
+			} else {
+				info->fps = 30.0f; // Default fallback
```

Comment:
If we don't know the FPS, maybe we should fail rather than play with the wrong fps.